neqo-http3/src/connection_client.rs

```diff
@@ -2364,4 +2430,676 @@ mod tests {
+
+        check_response(&mut client, request_stream_id);
+
+        let _out = server.process(out.dgram(), now());
+        assert_eq!(*server.state(), State::Connected);
+
+        assert!(client.tls_info().unwrap().resumed());
+        assert!(server.tls_info().unwrap().resumed());
+    }
+
+    #[test]
+    fn zero_rtt_before_resumption_token() {
+        let mut server = default_http3_client();
+        assert!(server
+            .fetch("GET", "https", "something.com", "/", &[])
+            .is_err());
+    }
+
+    #[test]
+    fn zero_rtt_send_reject() {
+        let (mut client, mut server) = connect();
+        let token = exchange_token(&mut client, &mut server.conn);
+
+        let mut client = default_http3_client();
+
+        // Using a freshly initialized anti-replay context
+        // should result in the server rejecting 0-RTT.
+        let ar = AntiReplay::new(now(), test_fixture::ANTI_REPLAY_WINDOW, 1, 3)
+            .expect("setup anti-replay");
+        let mut server = Connection::new_server(
+            test_fixture::DEFAULT_KEYS,
+            test_fixture::DEFAULT_ALPN,
+            &ar,
+            Rc::new(RefCell::new(FixedConnectionIdManager::new(10))),
+        )
+        .unwrap();
+
+        assert_eq!(client.state(), Http3State::Initializing);
+        client
+            .set_resumption_token(now(), &token)
+            .expect("Set resumption token.");
+
+        // Send ClientHello.
+        let client_hs = client.process(None, now());
+        assert!(client_hs.as_dgram_ref().is_some());
+
+        // Create a request
+        let request_stream_id = client
+            .fetch("GET", "https", "something.com", "/", &[])
+            .unwrap();
+        assert_eq!(request_stream_id, 0);
+
+        let client_0rtt = client.process(None, now());
+        assert!(client_0rtt.as_dgram_ref().is_some());
+
+        let server_hs = server.process(client_hs.dgram(), now());
+        assert!(server_hs.as_dgram_ref().is_some()); // Should produce ServerHello etc...
+        let server_ignored = server.process(client_0rtt.dgram(), now());
+        assert!(server_ignored.as_dgram_ref().is_none());
+
+        // The server shouldn't receive that 0-RTT data.
+        let recvd_stream_evt = |e| matches!(e, ConnectionEvent::NewStream { .. });
+        assert!(!server.events().any(recvd_stream_evt));
+
+        // Client should get a rejection.
+        let client_out = client.process(server_hs.dgram(), now());
+        assert!(client_out.as_dgram_ref().is_some());
+        let recvd_0rtt_reject = |e| e == Http3ClientEvent::ZeroRttRejected;
+        assert!(client.events().any(recvd_0rtt_reject));
+
+        // ...and the client stream should be gone.
+        let res = client.stream_close_send(request_stream_id);
+        assert!(res.is_err());
+        assert_eq!(res.unwrap_err(), Error::InvalidStreamId);
+
+        // Client will send Setting frame and open new qpack streams.
```

Comment:
This might make the test too complex, but it would be good to test now that the client can still make a request.  You might like to check that it sits on stream 0 too.